src/main.rs

```diff
@@ -1,42 +1,137 @@
-  let res = stylesheet.to_css(PrinterOptions { minify: true, ..PrinterOptions::default()}).unwrap();
+    let mut res = stylesheet
+        .to_css(PrinterOptions {
+            minify: cli_args.minify,
+            source_map: cli_args.sourcemap,
+            ..PrinterOptions::default()
+        })
+        .unwrap();
+
+    let map = if let Some(ref mut source_map) = res.source_map {
+        source_map.set_source_content(0, &res.code).map_err(|_| io::Error::new(io::ErrorKind::Other, "Error setting sourcemap"))?;
+        let mut vlq_output: Vec<u8> = Vec::new();
+        source_map.write_vlq(&mut vlq_output).map_err(|_| io::Error::new(io::ErrorKind::Other, "Error writing sourcemap vlq"))?;
+
+        let sm = SourceMapJson {
+            version: 3,
+            mappings: unsafe { String::from_utf8_unchecked(vlq_output) },
+            sources: source_map.get_sources(),
+            sources_content: source_map.get_sources_content(),
+            names: source_map.get_names()
+        };
+
+        serde_json::to_vec(&sm).ok()
+    } else {
+        None
+    };
+    if let Some(output_file) = &cli_args.output_file {
+        let mut code = res.code;
+        if cli_args.sourcemap {
+            if let Some(map_buf) = map {
+                let map_filename: String = output_file.to_owned() + ".map";
+                code += &format!("\n/*# sourceMappingURL={} */\n", map_filename);
+                fs::write(map_filename, map_buf)?;
+            }
+        }
+
+        fs::write(output_file, code.as_bytes())?;
 
-  if let Some(output_file) = cli_args.output_file {
-    fs::write(output_file, res.code.as_bytes())?;
-  } else {
-    println!("{}", res.code);
-  }
+        if cli_args.css_modules {
+            let css_modules_filename = cli_args
+                .css_modules_output_file
+                .unwrap_or(infer_css_modules_filename(cli_args.output_file)?);
+            if let Some(exports) = res.exports {
+                let css_modules_json = serde_json::to_string(&exports)?;
+                fs::write(css_modules_filename, css_modules_json)?;
+            }
+        }
 
-  Ok(())
+
+    } else {
+        println!("{}", res.code);
+        if cli_args.css_modules {
+            let css_modules_json = serde_json::to_string(&res.exports)?;
+            println!("{}", css_modules_json);
+        }
+    }
+
+    Ok(())
+}
+
+fn infer_css_modules_filename(output_file: Option<String>) -> Result<String, std::io::Error> {
+    if let Some(file) = output_file {
+        let path = path::Path::new(&file);
+        if path.extension() == Some(ffi::OsStr::new("json")) {
+            Err(io::Error::new(io::ErrorKind::Other, "Cannot infer a css modules json filename, since the output file extension is '.json'"))
+        } else {
+            // unwrap: the filename option is a String from clap, so is valid utf-8
+            Ok(path.with_extension("json").to_str().unwrap().into())
+        }
+    } else {
+        Ok("styles.json".into())
+    }
```

Comment:
on rereading this, I don't actually think we ever infer the filename if we don't have an output file specified; this should take a string and shouldn't have the outer condition